src/util/pup_stl.h

```diff
@@ -520,7 +502,7 @@ using Requires = typename requires_impl<
       // the shared ptr must be created with the original PUP::able ptr
       // otherwise the dynamic cast can lead to invalid free's
       // (it changes the pointer's address)
-      t = std::dynamic_pointer_cast<T>(std::shared_ptr<PUP::able>(_));
+      t = std::static_pointer_cast<T>(std::shared_ptr<PUP::able>(_));
```

Comment:
(We know this is a valid cast so we can use `static` instead of `dynamic` cast.)